Bottom: fa8bf236032a6911f02ffecf40da83d4f5171e73
Top:    78d76d841f4c64ec0108d514c53c6c6798564362
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-02 13:20:23 +0000

Push special case thindraw vs. fill polygon handling into the GUIs

This keeps the API more consistent between polygons and other object
types by introducing the new (HID_DRAW *)->draw_pcb_polygon() API.

I've left the (HID_DRAW *)->{fill,thindraw}_pcb_polygon() APIs for
now, as it means we can avoid a little code duplucation between the
GUIs. We may consider removing (or hiding) these APIs at some point,
as they are now _NOT_ intended to be called from outside of the
GUI's (HID_DRAW *)->draw_pcb_polygon() implementation.


---

diff --git a/src/draw.c b/src/draw.c
index d70963e..4a28844 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -799,25 +799,7 @@ poly_callback (const BoxType * b, void *cl)
 
   set_layer_object_color (i->layer, (AnyObjectType *) polygon);
 
-  if (gui->graphics->thindraw_pcb_polygon != NULL &&
-      (TEST_FLAG (THINDRAWFLAG, PCB) ||
-       TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
-    gui->graphics->thindraw_pcb_polygon (Output.fgGC, polygon, i->drawn_area);
-  else
-    gui->graphics->fill_pcb_polygon (Output.fgGC, polygon, i->drawn_area);
-
-  /* If checking planes, thin-draw any pieces which have been clipped away */
-  if (gui->graphics->thindraw_pcb_polygon != NULL &&
-      TEST_FLAG (CHECKPLANESFLAG, PCB) &&
-      !TEST_FLAG (FULLPOLYFLAG, polygon))
-    {
-      PolygonType poly = *polygon;
-
-      for (poly.Clipped = polygon->Clipped->f;
-           poly.Clipped != polygon->Clipped;
-           poly.Clipped = poly.Clipped->f)
-        gui->graphics->thindraw_pcb_polygon (Output.fgGC, &poly, i->drawn_area);
-    }
+  gui->graphics->draw_pcb_polygon (Output.fgGC, polygon, i->drawn_area);
 
   return 1;
 }
diff --git a/src/hid/common/draw_helpers.c b/src/hid/common/draw_helpers.c
index 02473be..9ea5094 100644
--- a/src/hid/common/draw_helpers.c
+++ b/src/hid/common/draw_helpers.c
@@ -257,6 +257,26 @@ should_compute_no_holes (PolygonType *poly, const BoxType *clip_box)
 #undef BOUNDS_INSIDE_CLIP_THRESHOLD
 
 void
+common_gui_draw_pcb_polygon (hidGC gc, PolygonType *polygon, const BoxType *clip_box)
+{
+  if (TEST_FLAG (THINDRAWFLAG, PCB) || TEST_FLAG (THINDRAWPOLYFLAG, PCB))
+    gui->graphics->thindraw_pcb_polygon (gc, polygon, clip_box);
+  else
+    gui->graphics->fill_pcb_polygon (gc, polygon, clip_box);
+
+  /* If checking planes, thin-draw any pieces which have been clipped away */
+  if (TEST_FLAG (CHECKPLANESFLAG, PCB) && !TEST_FLAG (FULLPOLYFLAG, polygon))
+    {
+      PolygonType poly = *polygon;
+
+      for (poly.Clipped = polygon->Clipped->f;
+           poly.Clipped != polygon->Clipped;
+           poly.Clipped = poly.Clipped->f)
+        gui->graphics->thindraw_pcb_polygon (gc, &poly, clip_box);
+    }
+}
+
+void
 common_fill_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box)
 {
   if (!poly->NoHolesValid)
@@ -589,6 +609,7 @@ common_draw_helpers_init (HID_DRAW *graphics)
   graphics->draw_pcb_line        = common_draw_pcb_line;
   graphics->draw_pcb_arc         = common_draw_pcb_arc;
   graphics->draw_pcb_text        = common_draw_pcb_text;
+  graphics->draw_pcb_polygon     = common_fill_pcb_polygon; /* Default is the non-GUI case */
 
   graphics->fill_pcb_polygon     = common_fill_pcb_polygon;
   graphics->thindraw_pcb_polygon = common_thindraw_pcb_polygon;
diff --git a/src/hid/common/draw_helpers.h b/src/hid/common/draw_helpers.h
index 95c2dc1..fb5730c 100644
--- a/src/hid/common/draw_helpers.h
+++ b/src/hid/common/draw_helpers.h
@@ -1,7 +1,6 @@
-void common_fill_pcb_polygon (hidGC gc, PolygonType *poly,
-                              const BoxType *clip_box);
-void common_thindraw_pcb_polygon (hidGC gc, PolygonType *poly,
-                                  const BoxType *clip_box);
+void common_gui_draw_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box);
+void common_fill_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box);
+void common_thindraw_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box);
 void common_fill_pcb_pad (hidGC gc, PadType *pad, bool clear, bool mask);
 void common_thindraw_pcb_pad (hidGC gc, PadType *pad, bool clear, bool mask);
 void common_fill_pcb_pv (hidGC fg_gc, hidGC bg_gc, PinType *pv, bool drawHole, bool mask);
diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index bab0bff..e9662d1 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -148,7 +148,7 @@ nogui_fill_polygon (hidGC gc, int n_coords, Coord *x, Coord *y)
 }
 
 static void
-nogui_fill_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box)
+nogui_draw_pcb_polygon (hidGC gc, PolygonType *poly, const BoxType *clip_box)
 {
   CRASH;
 }
@@ -501,7 +501,7 @@ common_nogui_graphics_init (HID_DRAW *graphics)
   graphics->fill_polygon =    nogui_fill_polygon;
   graphics->fill_rect =       nogui_fill_rect;
 
-  graphics->fill_pcb_polygon = nogui_fill_pcb_polygon;
+  graphics->draw_pcb_polygon = nogui_draw_pcb_polygon;
   graphics->fill_pcb_pad =     nogui_fill_pcb_pad;
   graphics->thindraw_pcb_pad = nogui_thindraw_pcb_pad;
   graphics->fill_pcb_pv =      nogui_fill_pcb_pv;
diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index f12fba3..8e39be4 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -2197,6 +2197,8 @@ hid_gtk_init ()
   ghid_graphics.fill_polygon        = ghid_fill_polygon;
   ghid_graphics.fill_rect           = ghid_fill_rect;
 
+  ghid_graphics.draw_pcb_polygon    = common_gui_draw_pcb_polygon;
+
   hid_register_hid (&ghid_hid);
 #include "gtk_lists.h"
 }
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index 3e2fa01..7caccee 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -4091,6 +4091,8 @@ hid_lesstif_init ()
   lesstif_hid.flush_debug_draw        = lesstif_flush_debug_draw;
   lesstif_hid.finish_debug_draw       = lesstif_finish_debug_draw;
 
+  lesstif_hid.draw_pcb_polygon        = common_gui_draw_pcb_polygon;
+
   lesstif_hid.graphics                = &lesstif_graphics;
 
   lesstif_graphics.make_gc             = lesstif_make_gc;
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index 597ff6d..9a95854 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -1244,7 +1244,7 @@ ps_fill_polygon (hidGC gc, int n_coords, Coord *x, Coord *y)
 }
 
 static void
-ps_fill_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
+ps_draw_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
 {
   /* Ignore clip_box, just draw everything */
 
@@ -1518,7 +1518,7 @@ void ps_ps_graphics_init (HID_DRAW *graphics)
   graphics->fill_polygon       = ps_fill_polygon;
   graphics->fill_rect          = ps_fill_rect;
 
-  graphics->fill_pcb_polygon   = ps_fill_pcb_polygon;
+  graphics->draw_pcb_polygon   = ps_draw_pcb_polygon;
 }
 
 void
diff --git a/src/hid_draw.h b/src/hid_draw.h
index c5fab7d..b65d804 100644
--- a/src/hid_draw.h
+++ b/src/hid_draw.h
@@ -54,6 +54,7 @@ struct hid_draw_st
   void (*draw_pcb_line) (hidGC gc, LineType *line);
   void (*draw_pcb_arc) (hidGC gc, ArcType *arc);
   void (*draw_pcb_text) (hidGC gc, TextType *, Coord);
+  void (*draw_pcb_polygon) (hidGC gc, PolygonType *poly, const BoxType *clip_box);
 
   void (*fill_pcb_polygon) (hidGC gc, PolygonType *poly, const BoxType *clip_box);
   void (*thindraw_pcb_polygon) (hidGC gc, PolygonType *poly, const BoxType *clip_box);
